### DISCUSSION TITLE
Test process improments

### DISCUSSION THREAD

**Discussion by HawtinZeng** [antvis/G6#6331] [General]
Recently, I commit some changes and ran tests in the repo, but it took me much time to run tests, wait, check svg diff, modify test code, it's annoying. So I start this discussion to list some actionable directions to improve the DX in test.
1. Compare svg rendering result instead of svg strings.
2. Run one a suite ( a .spec.ts  file) instead of all suits.
3. Simplify the test output, remove all warnings.
4. Add pre-push hook to run test and check test passed or not.

**Comment by Aarebecca** [comment]
Thank you for your feedback. I'm sorry that our contribution guidelines are not yet as thorough as they should be. We're working on improving them. As for the points you raised:

1. Two years ago, we experimented with comparing visual results instead of strings, using tools like node-canvas or other SSR techniques to generate screenshots. However, this approach consumed a lot of storage space and the comparison accuracy wasn't reliable—sometimes two screenshots looked identical but it was purely coincidental, which led to undetected issues.
2. You can already run a single test suite using a command like `npx jest /path/to/the-test-suite.spec.ts`.
3. Our tests are based on Jest, and we haven't made significant adjustments to its default behavior.
4. Jest already supports pre-push hooks for running tests.

**Reply by HawtinZeng** [reply]
1. 可以保存svg，但是执行比较逻辑的时候将svg转换成位图数据，根据位图数据进行比较。位图数据没有具体的元素层次、结构，比较宽松一点，我想的是结合两种比较方式，对于一些关于层级、结构的场景使用svg比较，对于一些绘图效果相关的场景，使用位图比较。当然这会增加维护的工作量，需要看后面的场景需不需要。
2. 对于jest命令，我昨天试过了，好像不行，刚刚也试了下调了一下，确实可以，但是windows环境使用这个命令，需要用手动敲路径，使用`/`分割，不能用shell的自动补全生成的路径。
3. 这个有空我改改。
4. 好像没有跑，我在shell里面没有看到pre-push执行测试的过程。
